### ISSUE TITLE
Is it good to add short escaping alias \s for space (\u0020)?

### ISSUE THREAD

**Issue by LongTengDao** [toml-lang/toml#622]
To avoid spaces at line end in multi-line string being auto trimmed accidentally, and simplify `\u0020` writting.

**Comment by eksortso** [comment]
I'm a little fond of this idea. It makes spaces explicit. But it's definitely new syntax.

And mentally, the notion conflicts with the regular-expression character class "`\s`", which is used to match any sort of whitespace, and not just `\u0020`. Different use cases for these, of course.

Is there a formulation of this bit of syntax that doesn't conflict with past usages anywhere?

**Comment by LongTengDao** [comment]
Or define a line end wrapper character which not output?

```
a = """
There are\s
two ways \# even comment
to write space at line end."""
```

```json
{ "a": "There are \ntwo ways \nto write space at line end." }
```

**Comment by pradyunsg** [comment]
Marking as post-1.0 since I don't really have the bandwidth to think about the edge cases here and this isn't super important to get to 1.0.

**Comment by LongTengDao** [comment]
> Marking as post-1.0 since I don't really have the bandwidth to think about the edge cases here and this isn't super important to get to 1.0.

Sure, focus on getting 1.0 things done first, best wishes!

**Comment by ChristianSi** [comment]
Even if this is post-1.0, it was supposedly meant to stay open until then?

**Comment by eksortso** [comment]
@LongTengDao, since this issue is already tagged post-1.0, could you please reopen it? We can effectively ignore it until after 1.0 is released, as I think the issue still bears some examination. You could even argue to remove that tag if you think it's merited.

**Comment by pradyunsg** [comment]
I can reopen it now but if it's closed again by OP, I won't be able to reopen.

**Comment by mmakaay** [comment]
Quite confusing if I use regexps in my toml docs (which I do), and having to escape \s as \\s. For other escaped chars it's kind of natural to do so since they are generally used well known escapes. The \s to represent a space would be rather unique to TOML (as far as my knowledge goes), which might cause people to trip over it.

```
my.regexp = "something\\sfeels\sfishy"
```

**Comment by ChristianSi** [comment]
@mmakaay: For regexes, literal strings might, in any case, be a better fit.

**Comment by LongTengDao** [comment]
@ChristianSi @eksortso @pradyunsg Sorry, I see.

@mmakaay `"\s"` is also invalid now in TOML, you need always write `"\\s"` or `'\s'`.

But the different meaning of `\s` in regExp is a problem indeed.

**Comment by pradyunsg** [comment]
Let's defer further discussion until after TOML 1.0?

---

Noting a summary for self-reference later.

This is a request for an additional compact escape sequence: \s -> U+0020 (space), for easier trailing spaces in multiline strings when trailing spaces are auto-trimmed. It is already possible to denote this as `\x0020`.

The question to answer here is -- is there enough benefit to adding a compact notation for `\x0020` for the specified use case? I'm intuitively leaning "no" but have decided on anything yet, since some time should be spent thinking about all the details/implications.

**Comment by eksortso** [comment]
> `\x0020`

You mean `\u0020`?

**Comment by YoYoYonnY** [comment]
Why not just `\ ` (Literally escaping a space character), it is less confusing.
By the way, I also like Lua's `\z` escape, which does something similair (Remove all following space characters including newlines).

**Comment by eksortso** [comment]
A backslash followed by a space is not very visible. I think the point of having the code was to make the character stand out.

We could make `\x` a valid prefix for two-character codes, for bytes and ASCII (and ASCII code points in UTF-8 for that matter). A space would be `\x20`, for instance.

Granted that's not as small as the proposed two-character `\s`, but it'd be more useful. After all, I think the README uses "\x" in places, doesn't it?

**Comment by abelbraaksma** [comment]
I'm not so fond of `\s`, because in a lot of languages, and specifically in regular expressions, this already has a special meaning. Currently, using it raises an error. Once that's gone, it'll be confusing to people.

I like the `\xXX` idea, though. It even just `\uXX`, shouldn't be hard to parse, and helps the generic case of escaping an ISO-8859-1 character.

**Comment by LongTengDao** [comment]
> Why not just `\ ` (Literally escaping a space character), it is less confusing.
> By the way, I also like Lua's `\z` escape, which does something similair (Remove all following space characters including newlines).

\s （或其它方案）的目的是避免行末空白被半智能文本编辑器意外清除，换言之是为了保留空白，而不是为了删除空白。因此\ 的方案无法做到这点，尾空白还是有可能被意外删除。另外，TOML中已经有了\ 的设计，它就是用来删除后续空白的。

The purpose of `\s` (or other solution) is to prevent eol whitespace from being accidentally cleared by a semi-intelligent text editor, in other words to preserve whitespace, not to remove it. Therefore, `\ ` cannot do this, the tail space may still be accidentally deleted. BTW: there is already a design about `\ ` in TOML for removing subsequent whitespace as you wish.

**Comment by marzer** [comment]
@eksortso I second the `\xXX` syntax. It's already used in other languages that _also_ allow `\uXXXX` and `\UXXXXXXXX`, so the consistency is welcome.

**Comment by abelbraaksma** [comment]
> BTW: there is already a design about \ in TOML for removing subsequent whitespace as you wish.

That's true. And doesn't that already solve the original issue? If you're editor is over zealous in removing tailing white space, you can do this:

```toml
key = """
   Alea \
   iacta \
   est""" 
```

**Comment by eksortso** [comment]
> That's true. And doesn't that already solve the original issue?

Not really. The end-of-line backslash strips away _all_ whitespace and _all_ newlines following it before the next non-space, non-linebreak character. That means that subsequent line breaks and indentation at the beginning of the following line gets stripped by the `\` too. We're looking for an alternative that doesn't do that.

Let's take your example:
```toml
key = """
   Alea \
   iacta \
   est"""
```

This is equivalent to the one-liner `key = "Alea iacta est"`. But what we want is for `key` to be set to this, with the three-space indentation on all three lines, and the one extra space at the end of the first two lines:
```
   Alea 
   iacta 
   est
```

(Of course, you'll need to highlight that or look at the source to notice the end-of-line spaces.)

One way to add a space prominently is with `\u0020`.

```toml
key = """
   Alea\u0020
   iacta\u0020
   est"""
```

If I can get a PR together later today, then `\x20` will work in some post-v1.0.0 version of TOML.

**Comment by abelbraaksma** [comment]
@eksortso, thanks, I didn't think I'd that. I typically wasn't too get rid of extraneous whitespace, but I guess there are use cases (like code sections), where that isn't desirable. Still, wanting a space right before a newline seems pretty fringe to me.

Nonetheless, short escape codes have a wider use than just spaces, I'd definitely love having that ;)

**Comment by pradyunsg** [comment]
Let's add a `\xHH` -- I don't think a `\s` is particularly valuable once that's added in.

**Comment by pradyunsg** [comment]
Closing this since the answer to the original question: No. :)

#796 can stand on its own, or we can file a dedicated issue for it if it needs extended discussion.
